common/txmgr/address_state.go

```diff
@@ -0,0 +1,387 @@
+	as.lock.Lock()
+	defer as.lock.Unlock()
+
+	as.unstarted.Close()
+	as.unstarted = nil
+	as.inprogress = nil
+	clear(as.unconfirmed)
+	clear(as.idempotencyKeyToTx)
+}
+
+func (as *AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) unstartedCount() int {
+	as.lock.RLock()
+	defer as.lock.RUnlock()
+
+	return as.unstarted.Len()
+}
+func (as *AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) unconfirmedCount() int {
+	as.lock.RLock()
+	defer as.lock.RUnlock()
+
+	return len(as.unconfirmed)
+}
+
+func (as *AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) findTxWithIdempotencyKey(key string) *txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE] {
+	as.lock.RLock()
+	defer as.lock.RUnlock()
+
+	return as.idempotencyKeyToTx[key]
+}
+
+func (as *AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) findLatestSequence() SEQ {
+	as.lock.RLock()
+	defer as.lock.RUnlock()
+
+	var maxSeq SEQ
+	if as.inprogress != nil && as.inprogress.Sequence != nil {
+		if (*as.inprogress.Sequence).Int64() > maxSeq.Int64() {
+			maxSeq = *as.inprogress.Sequence
+		}
+	}
+	for _, tx := range as.unconfirmed {
+		if tx.Sequence == nil {
+			continue
+		}
+		if (*tx.Sequence).Int64() > maxSeq.Int64() {
+			maxSeq = *tx.Sequence
+		}
+	}
+
+	return maxSeq
+}
+
+func (as *AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) peekNextUnstartedTx() (*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], error) {
+	as.lock.RLock()
+	defer as.lock.RUnlock()
+
+	tx := as.unstarted.PeekNextTx()
+	if tx == nil {
+		return nil, fmt.Errorf("peek_next_unstarted_tx: %w (address: %s)", ErrTxnNotFound, as.fromAddress)
+	}
+
+	return tx, nil
+}
+
+func (as *AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) peekInProgressTx() (*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], error) {
+	as.lock.RLock()
+	defer as.lock.RUnlock()
+
+	tx := as.inprogress
+	if tx == nil {
+		return nil, fmt.Errorf("peek_in_progress_tx: %w (address: %s)", ErrTxnNotFound, as.fromAddress)
+	}
+
+	return tx, nil
+}
+
+func (as *AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) addTxToUnstarted(tx *txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]) error {
+	as.lock.Lock()
+	defer as.lock.Unlock()
+
+	if as.unstarted.Len() >= as.unstarted.Cap() {
+		return fmt.Errorf("move_tx_to_unstarted: address %s unstarted queue capactiry has been reached", as.fromAddress)
+	}
+
+	as.unstarted.AddTx(tx)
+	if tx.IdempotencyKey != nil {
+		as.idempotencyKeyToTx[*tx.IdempotencyKey] = tx
+	}
+
+	return nil
+}
+
+func (as *AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) moveUnstartedToInProgress(tx *txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]) error {
+	as.lock.Lock()
+	defer as.lock.Unlock()
+
+	if as.inprogress != nil {
+		return fmt.Errorf("move_unstarted_to_in_progress: address %s already has a transaction in progress", as.fromAddress)
+	}
+
+	if tx != nil {
```

Comment:
moveUnstartedToInProgress accepts tx as argument, but allows it to be nil. In that case we get tx from the priority queue. In which cases caller of the `moveUnstartedToInProgress` might want to use their own value of tx instead of one from the queue?